src/Forms/GridField/GridFieldDetailForm_ItemRequest.php

```diff
@@ -561,37 +561,115 @@ public function getEditLink($id)
+        return $list->limit($limit, $limitOffset)->column('ID');
+    }
+
+    /**
+     * Get the current paginator state
+     */
+    private function getGridFieldPaginatorState(): GridState_Data
+    {
+        $state = $this->getGridField()->getState(false);
+        $gridStateStr = $this->getStateManager()->getStateFromRequest($this->gridField, $this->getRequest());
+        if (!empty($gridStateStr)) {
+            $state->setValue($gridStateStr);
+        }
+
+        return $state->getData()->getData('GridFieldPaginator');
+    }
+
+    /**
+     * Get the grid state for an adjacent record
+     */
+    private function getGridStateForAdjacentRecord(int $offset): GridState_Data
+    {
+        $gridField = $this->getGridField();
+        $map = $this->getGridFieldItemAdjacencies();
+        if (empty($map)) {
+            throw new LogicException('No adjacent records exist');
+        }
+
+        $state = clone $gridField->getState();
+        $index = array_search($this->record->ID, $map);
+        $position = $index + $offset;
+
+        $currentPage = $this->getGridFieldPaginatorState()->getData('currentPage');
+        $itemsPerPage = $this->getGridFieldPaginatorState()->getData('itemsPerPage');
+        $page = $currentPage;
+        $hasMorePages = $this->getNumPages($gridField) > $currentPage;
+
+        if ($position === 0 && $currentPage > 1) {
+            $page = $currentPage - 1;
+        } elseif ($hasMorePages && $position >= $itemsPerPage + 1) {
+            $page = $currentPage + 1;
+        }
+        $state->GridFieldPaginator->currentPage = (int)$page;
+
+        return $state;
+    }
+
+    /**
+     * Get the edit link for an adjacent record
+     */
+    private function getEditLinkForAdjacentRecord(int $offset): string
+    {
+        $link = Controller::join_links(
+            $this->gridField->Link(),
+            'item',
+            $this->getAdjacentRecordID($offset)
+        );
+        $state = $this->getGridStateForAdjacentRecord($offset);
+        // Get a dummy gridfield so we can set some future state without affecting the current gridfield
+        $gridField = clone $this->gridField;
+        $gridField->getState(false)->setValue($state);
+        return $gridField->addAllStateToUrl($link);
+    }
+
+    /**
+     * @param int $offset The offset from the current record
+     * @return int|bool
+     */
+    private function getAdjacentRecordID($offset)
+    {
+        $map = $this->getGridFieldItemAdjacencies();
+        if (empty($map)) {
+            return false;
+        }
         $index = array_search($this->record->ID, $map ?? []);
-        return isset($map[$index+$offset]) ? $map[$index+$offset] : false;
+        $position = $index + $offset;
+        return isset($map[$position]) ? $map[$position] : false;
+    }
+
+    /**
+     * Gets the number of GridField pages
+     */
+    public function getNumPages(GridField $gridField): int
```

Comment:
```suggestion
    private function getNumPages(GridField $gridField): int
```
As per https://github.com/silverstripe/silverstripe-framework/pull/10439#discussion_r941884395

Reply:
DONE